app/app.js

```diff
@@ -6,69 +6,100 @@ window.run = function() {
+        var errorControllerPromise = ErrorController.init();
+        var welcomeModalControllerPromise = WelcomeModalController.init(errorControllerPromise);
+        var cartModalControllerPromise = CartModalController.init(errorControllerPromise);
         var cartEventHandlerPromise = cartModalControllerPromise.then(
             function(cartModalController) {
                 return function() {
                     cartModalController.show();
                 };
             });
-
+        // Register RPC to expose whether it is possible to go back.
+        // This is necessary to determine whether to show/hide
+        // the back button on the error modal
+        var registerCanGoBackRpc = function(controller) {
+            Promise.join(welcomeModalControllerPromise, cartModalControllerPromise,
+            function(welcomeModal, cartModal) {
+                Astro.registerRpcMethod(AstroRpc.names.appCanGoBack, [], function(res) {
+                    // We want to return true when the cart is showing since
+                    // calling `back` is equivalent to dismissing the cart
+                    if (cartModal.isShowing) {
+                        res.send(null, true);
+                    } else if (welcomeModal.isShowing) {
+                        welcomeModal.canGoBack().then(function(canGoBack) {
+                            res.send(null, canGoBack);
+                        });
+                    } else {
+                        controller.canGoBack().then(function(canGoBack) {
+                            res.send(null, canGoBack);
+                        });
+                    }
+                });
+            });
+        };
         var createTabBarLayout = function(counterBadgeControllerPromise) {
             var layoutPromise = AnchoredLayoutPlugin.init();
             var tabBarControllerPromise = TabBarController.init(
-                layoutPromise, cartEventHandlerPromise, counterBadgeControllerPromise);
+                    layoutPromise,
+                    cartEventHandlerPromise,
+                    counterBadgeControllerPromise,
+                    errorControllerPromise
+                );
 
             var layoutSetupPromise = Promise.join(
                 layoutPromise,
                 tabBarControllerPromise,
             function(layout, tabBarController) {
                 layout.addBottomView(tabBarController.tabBar);
+                Application.setMainViewPlugin(layout);
 
-                return Application.setMainViewPlugin(layout);
-            });
-
-            // Tab layout must be added as the mainViewPlugin before
-            // The first tab is selected or else the navigation does
-            // not complete correctly
-            return Promise.join(tabBarControllerPromise, layoutSetupPromise,
-            function(tabBarController) {
+                return tabBarController;
+            }).then(function(tabBarController) {
```

Comment:
Do we need to return the `tabBarController` here and then use `.then()` right next? It seems like we could eliminate these two lines and just call `selectTab` right away.